### PR TITLE
Improve MsgOfInterest tags, add codec documentation

### PR DESCRIPTION
**Pull request by Rqnsom**
Two commits:
```git
refactor: change MsgOfInterest tags to Hashset

This is a minor enhancement/refactoring:
- MsgOfIntrest tags: replace string to HashSet
- rename "websocketcodec" file to "websocket" (align with convention)
```
```git
doc: include codec code documentation
```

### CODE REVIEW

**Comment by Rqnsom** [comment]
I tried to include this fix as well:
https://github.com/runziggurat/algorand/pull/16#pullrequestreview-1165405457

but it doesn't work because span does not implement the Default trait.
```
        Self {
            span,
            tag: None,
            topic: TopicCodec::default(),
        }
```

File: src/protocol/codecs/topic.rs [hunk 1]
```diff
@@ -28,9 +30,17 @@ impl MsgOfInterest {
             return Err(invalid_data!("expected 'tags' topic"));
         }
 
-        Ok(Self {
-            tags: tag_topic.value,
-        })
+        let tags = tag_topic.value.split(',');
+        let expected_len = tags.clone().count();
```
Comment (lwojtow) [inline src/protocol/codecs/topic.rs:34]: Do we need this `clone` here?

**Review by lwojtow — APPROVED** [review]

File: src/protocol/codecs/topic.rs (hunk 1 above)
Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:34]: We do need to copy it because `count()` consumes the iterator

Comment (lwojtow) [reply src/protocol/codecs/topic.rs:34]: IDK how large `tags` usually is, or how performance critical this path is, but how about:
`let expected_len = 1 + tag_topic.value.chars().filter(|c| *c == ',').count();`
On my machine that's like 30% quicker.

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:34]: Aha, good to know!
The tags are pretty short here. Here's an example of what we receive from the node:
`UE,NP,SP,TS,MI,TX,UC,MS,pi,AV,pj,PP,VB`

Do you think we should apply your suggestion here?
The solution you find more readable is the solution I vote for 🤔

Comment (lwojtow) [reply src/protocol/codecs/topic.rs:34]: Actually, I think my suggestion is less readable, but 30% performance hit is not something I give up easily. IDK how hot this path is, so your call :)

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:34]: We receive this message only once after the handshake, so the path is quite a cold one 🧊

File: src/protocol/codecs/topic.rs [hunk 2]
```diff
@@ -28,9 +30,17 @@ impl MsgOfInterest {
...
+        let tags = tags
+            .map_while(|tag| Tag::try_from(tag).ok())
+            .collect::<HashSet<Tag>>();
+
+        if tags.len() != expected_len {
+            return Err(invalid_data!("invalid tag received"));
+        }
```
Comment (ljedrz) [inline src/protocol/codecs/topic.rs:41]: this is fine, but I'd instead go with the following:
```
let tags = tags
            .map(Tag::try_from)
            .collect::<io::Result<HashSet<Tag>>>()?;
```
there is no need to check the number of commas then, I think

Comment (ljedrz) [reply src/protocol/codecs/topic.rs:41]: alternatively `let tags: HashSet<_> = (...)` (without the explicit type by `collect`), which should work too, but is shorter

**Review by ljedrz — APPROVED** [review]
LGTM with one suggestion

File: src/protocol/codecs/topic.rs (hunk 2 above)
Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:41]: Great! I was trying to find a solution like that but couldn't make it work... 😅
Thanks!

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:41]: Actually, collect still expects some type annotations:

This is a compromise:
```rust
        let tags: HashSet<_> = tag_topic
            .value
            .split(',')
            .map(Tag::try_from)
            .collect::<io::Result<_>>()?;
```

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs]: Done

**Comment by ljedrz** [comment]
as for this:
> I tried to include this fix as well:
https://github.com/runziggurat/algorand/pull/16#pullrequestreview-1165405457
but it doesn't work because span does not implement the Default trait.

I was suggesting this:
```
Self {
    span,
    ..Default::default()
}
```
which doesn't require the `Span` to impl `Default`

**Pull request merged**